### PR TITLE
PCX-500: Update error handling

### PR DESCRIPTION
**Pull request by siebeprojects**
Internal SDK Errors are now always combined with an Interaction object.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -198,9 +198,7 @@ Code sample how to initialize and display the Payment Page:
 Payment Result
 ==============
 
-Payment results are returned through the onActivityResult() method in your Activity. When the page is closed, the returned PaymentResult class contains information about the performed operation. I.e. it may contain an Interaction and OperationResult object describing the state of the latest Charge operation.
-
-The Interaction and OperationResult objects are never created by the Android SDK but instead come from the Optile Payment API. The PaymentError object inside the PaymentResult class is created by the Android SDK and contains information about an error that happened inside the Android SDK. 
+Payment results are returned through the onActivityResult() method in your Activity. When the payment page is closed, the returned PaymentResult class contains information about the operation request. I.e. it may contain an Interaction and OperationResult describing the state of a Charge operation. The PaymentError object inside the PaymentResult class is created when an error occurred inside the Android-SDK.
```
Comment (hjhill) [inline README.rst:201]: "PaymentResult class" looks wrong - do you mean "PaymentResult object"? The structure of the sentence looks a bit weird, the part "information about the operation request" is probably only understandable for optile insiders, and it remains unclear which subobjects of `PaymentResult` are optional and/or mandatory...

Idea for a better description: "When the payment page is closed, the `Intent` from its activity result can be converted to a `PaymentResult` containing (add list here, with clear distinction of what to expect in which case)"

File: README.rst [hunk 2]
```diff
@@ -218,47 +216,68 @@ Code sample how to obtain the PaymentResult inside the onActivityResult() method
         }
         String resultInfo = result.getResultInfo();
 
-        // Operation request has been made and "result" contains
-        // an Interaction and optional OperationResult describing the operation result
+        // Operation request has been made and "result" contains a resultInfo, an optional Interaction and OperationResult object.
+        // The Interaction object is null when the user selected the same PresetAccount again or closed the page. 
         if (resultCode == PaymentUI.RESULT_CODE_OK) {
```
Comment (hjhill) [inline README.rst:221]: Where does `resultCode` come from?

File: README.rst [hunk 3]
```diff
@@ -218,47 +216,68 @@ Code sample how to obtain the PaymentResult inside the onActivityResult() method
...
 Canceled
 ---------
 
-The RESULT_CODE_CANCELED code indicates that the Android SDK did not perform a successful operation. This may happen for different reasons, i.e. the user clicked the back button. The PaymentResult may contain an Interaction and an OperationResult with details about the failed operation. If both Interaction and OperationResult are null then the user closed the page before any request was made. 
-    
-Error
------
+The RESULT_CODE_CANCELED code indicates that the Android SDK did not perform a successful operation. The PaymentResult will always contain an Interaction and optional OperationResult or PaymentError indicating what went wrong and what should be done next. 
+
+1. When the code of the Interaction is VERIFY, it means that the Android-SDK was not able to determine the current state of the list and should be verified before continuing.
```
Comment (hjhill) [inline README.rst:248]: Something's missing here... what exactly should be verified?

File: README.rst [hunk 4]
```diff
@@ -218,47 +216,68 @@ Code sample how to obtain the PaymentResult inside the onActivityResult() method
...
+
+2. The user performed an operation and the result contains an OperationResult containing detailed information why the operation failed.
```
Comment (hjhill) [inline README.rst:250]: Maybe two sentences?
```suggestion
2. The user performed an operation that failed. The result contains an OperationResult with detailed information about the cause of the failure.
```

File: README.rst [hunk 5]
```diff
@@ -218,47 +216,68 @@ Code sample how to obtain the PaymentResult inside the onActivityResult() method
...
+
+3. An error occurred inside the Android-SDK and the result contains a PaymentError. When the PaymentResult contains a PaymentError, the PaymentResult will also contain a client-side created Interaction object. The following paragraph describes which types of Interactions are created by the Android-SDK when an internal error occurred.
```
Comment (hjhill) [inline README.rst:252]: Same here - I would prefer a very concise first sentence:
```suggestion
3. An error occurred inside the Android-SDK. In this case the PaymentResult contains a PaymentError and a client-side created Interaction object. The following paragraph describes which types of Interactions are created by the Android-SDK when an internal error occurred.
```

File: README.rst (hunk 3 above)
Comment (hjhill) [inline README.rst:248]: More generally about this list: can we define more clearly how the three reasons would be determined? The conditions you give here don't look like they cleanly separate between those three reasons.

File: README.rst [hunk 6]
```diff
@@ -218,47 +216,68 @@ Code sample how to obtain the PaymentResult inside the onActivityResult() method
...
++==================+=======================+=================================================================+
+| ABORT            | CLIENTSIDE_ERROR      | An internal error occurred inside the Android-SDK, i.e. a       |
+|                  |                       | SecurityException was thrown. The list may still be valid.      |      
++------------------+-----------------------+-----------------------------------------------------------------+
+| ABORT            | COMMUNICATION_FAILURE | A network failure occurred while communicating with the         |            
+|                  |                       | Optile Payment API. The list may still be valid.                |
++------------------+-----------------------+-----------------------------------------------------------------+
+| VERIFY           | CLIENTSIDE_ERROR      | An error occurred during a Charge operation.                    |
+|                  |                       | The charge may have been successful, therefor the status of the | 
+|                  |                       | payment (list) must be verified.                           |
```
Comment (hjhill) [inline README.rst:271]: Minor: missing spaces before the final character

File: README.rst [hunk 7]
```diff
@@ -218,47 +216,68 @@ Code sample how to obtain the PaymentResult inside the onActivityResult() method
...
++------------------+-----------------------+-----------------------------------------------------------------+
+| InteractionCode  | InteractionReason     | Description                                                     |
++==================+=======================+=================================================================+
+| ABORT            | CLIENTSIDE_ERROR      | An internal error occurred inside the Android-SDK, i.e. a       |
+|                  |                       | SecurityException was thrown. The list may still be valid.      |      
++------------------+-----------------------+-----------------------------------------------------------------+
+| ABORT            | COMMUNICATION_FAILURE | A network failure occurred while communicating with the         |            
+|                  |                       | Optile Payment API. The list may still be valid.                |
++------------------+-----------------------+-----------------------------------------------------------------+
+| VERIFY           | CLIENTSIDE_ERROR      | An error occurred during a Charge operation.                    |
```
Comment (hjhill) [inline README.rst:269]: So "VERIFY" belongs to case 3 (interaction created on the client-side) and also to case 1 (VERIFY) from the list above? Hmm... something does not add up here.

File: README.rst (hunk 2 above)
Comment (hjhill) [reply README.rst:221]: Sorry, I missed the paramter because I was a bit confused by `result` and `resultInfo`... maybe the code example could be made a bit more focused by introducing a `private void handlePaymentActivityResult(int resultCode, PaymentResult result)`?

File: README.rst [hunk 8]
```diff
@@ -218,47 +216,68 @@ Code sample how to obtain the PaymentResult inside the onActivityResult() method
...
-        if (resultCode == PaymentUI.RESULT_CODE_ERROR) {
             PaymentError error = result.getPaymentError();
         }
     }
 
 Successful
 ----------
 
-The RESULT_CODE_OK code indicates that the operation request was successful, there are two situations when this result is returned:
+The RESULT_CODE_OK code indicates that the operation request was successful, there are three situations when this result is returned:
```
Comment (hjhill) [inline README.rst:237]: Plus/minus one problem? :trollface:

File: example-demo/src/main/java/net/optile/example/demo/checkout/CheckoutPresenter.java [hunk 9]
```diff
@@ -70,8 +66,21 @@ private void handlePaymentSuccess(PaymentResult result) {
...
         Interaction interaction = result.getInteraction();
-        if (interaction != null && interaction.getCode() == InteractionCode.ABORT) {
-            view.closePayment();
+        if (interaction == null) {
+            return;
+        }
+        switch (interaction.getCode()) {
+            case InteractionCode.ABORT:
+                // When hasNetworkFailure() is true then the interaction reason is COMMUNICATION_FAILURE
+                if (result.hasNetworkFailureError()) {
```
Comment (hjhill) [inline example-demo/src/main/java/net/optile/example/demo/checkout/CheckoutPresenter.java:75]: `hasNetworkFailureError` looks like a pleonasm - `hasNetworkFailure()` looks better (I have even thought about `isNetworkFailure()` - not sure if it fits though)

File: example-demo/src/main/java/net/optile/example/demo/checkout/CheckoutPresenter.java [hunk 10]
```diff
@@ -70,8 +66,21 @@ private void handlePaymentSuccess(PaymentResult result) {
...
     private void handlePaymentCanceled(PaymentResult result) {
         Interaction interaction = result.getInteraction();
-        if (interaction != null && interaction.getCode() == InteractionCode.ABORT) {
-            view.closePayment();
+        if (interaction == null) {
+            return;
+        }
+        switch (interaction.getCode()) {
+            case InteractionCode.ABORT:
+                // When hasNetworkFailure() is true then the interaction reason is COMMUNICATION_FAILURE
```
Comment (hjhill) [inline example-demo/src/main/java/net/optile/example/demo/checkout/CheckoutPresenter.java:74]: How would that information help a potential implementer? (probably I just don't see the reason)

File: example-demo/src/main/java/net/optile/example/demo/summary/SummaryPresenter.java [hunk 11]
```diff
@@ -125,17 +121,26 @@ private void handlePaymentResult(SdkResult result) {
...
         Interaction interaction = result.getInteraction();
-        if (interaction != null && interaction.getCode() == InteractionCode.ABORT) {
-            view.closePayment(null);
+        if (interaction == null) {
+            return;
+        }
+        switch (interaction.getCode()) {
+            case InteractionCode.ABORT:
+                // When hasNetworkFailure() is true then the interaction reason is COMMUNICATION_FAILURE
+                if (result.hasNetworkFailureError()) {
```
Comment (hjhill) [inline example-demo/src/main/java/net/optile/example/demo/summary/SummaryPresenter.java:135]: same as in previous class

File: payment/src/main/java/net/optile/network/basic/BasicNetworkService.java [hunk 12]
```diff
@@ -90,41 +102,71 @@ public void onRedirectCanceled() {
...
+                presenter.onProcessPaymentResult(PaymentUI.RESULT_CODE_OK, result);
+        }
+    }
+
+    private void handleProcessPaymentError(PaymentError error) {
+        ErrorInfo errorInfo = error.getErrorInfo();
+        PaymentResult result;
+
+        if (errorInfo == null) {
+            result = new PaymentResult(errorInfo.getInteraction(), error);
```
Comment (hjhill) [inline payment/src/main/java/net/optile/network/basic/BasicNetworkService.java:151]: `errorInfo` is `null` here.

File: payment/src/main/java/net/optile/network/basic/BasicNetworkService.java [hunk 13]
```diff
@@ -90,41 +102,71 @@ public void onRedirectCanceled() {
...
+        } else {
+            String code = getErrorInteractionCode(operation);
+            String reason = error.getNetworkFailure() ? COMMUNICATION_FAILURE : CLIENTSIDE_ERROR;
+            Interaction interaction = new Interaction(code, reason);
+            result = new PaymentResult(interaction, error);
+        }
+        presenter.onProcessPaymentResult(PaymentUI.RESULT_CODE_CANCELED, result);
```
Comment (hjhill) [inline payment/src/main/java/net/optile/network/basic/BasicNetworkService.java:158]: Actually this method could be split to a one-liner:
```
presenter.onProcessPaymentResult(PaymentUI.RESULT_CODE_CANCELED, toPaymentResult(error));
```
and a method `private PaymentResult toPaymentResult(PaymentError error)` that converts the error.

File: payment/src/main/java/net/optile/payment/core/PaymentError.java [hunk 14]
```diff
@@ -8,74 +8,199 @@
...
+            Log.w("pay", e);
+            throw new RuntimeException(e);
+        }
+        this.cause = (Throwable) in.readSerializable();
+        this.networkFailure = in.readInt() == 1;
+    }
+
+
+    /**
+     * Helper class to construct a new InternalError from the given Throwable
```
Comment (hjhill) [inline payment/src/main/java/net/optile/payment/core/PaymentError.java:168]:
```suggestion
     * Constructs a new PaymentError from the given Throwable
```

File: payment/src/main/java/net/optile/payment/core/PaymentError.java [hunk 15]
```diff
@@ -8,74 +8,199 @@
...
+
+    public ErrorInfo getErrorInfo() {
+        return errorInfo;
+    }
+
+    public Throwable getCause() {
+        return cause;
+    }
+
+    public boolean getNetworkFailure() {
```
Comment (hjhill) [inline payment/src/main/java/net/optile/payment/core/PaymentError.java:139]: Maybe `isNetworkFailure()`? (proper Bean-style naming for `boolean`s)

File: payment/src/main/java/net/optile/payment/network/BaseConnection.java [hunk 16]
```diff
@@ -263,41 +263,45 @@ void writeToOutputStream(final HttpURLConnection conn, String data) throws IOExc
...
 
             if (!TextUtils.isEmpty(data) && !TextUtils.isEmpty(contentType) && contentType.contains(CONTENTTYPE_JSON)) {
-                info = gson.fromJson(data, ErrorInfo.class);
+                errorInfo = gson.fromJson(data, ErrorInfo.class);
             }
         } catch (IOException | JsonParseException e) {
             // Ignore the exceptions since the ErrorInfo is an optional field
             // and it is more important to not lose the status error code
-            Log.w("pay_BaseConnection", e);
+            Log.w("androidsdk", e);
```
Comment (hjhill) [inline payment/src/main/java/net/optile/payment/network/BaseConnection.java:284]: This tag looks a bit unspecific - do we want use something a bit more distinctive? (just think about what happens if a merchant uses different SDKs for his app...)

File: payment/src/main/java/net/optile/payment/ui/page/ChargePaymentPresenter.java [hunk 17]
```diff
@@ -127,19 +128,21 @@ public void onPaymentSessionSuccess(PaymentSession session) {
...
-        PaymentResult result = PaymentResult.fromThrowable(cause);
-        if (result.hasError()) {
-            handleLoadingError(result);
-        } else {
-            closeWithCanceledCode(result);
-        }
+        handleLoadingError(cause);
+    }
+
+    private PaymentResult createPaymentResult(PaymentError error) {
```
Comment (hjhill) [inline payment/src/main/java/net/optile/payment/ui/page/ChargePaymentPresenter.java:134]: The order of methods makes it hard to follow the flow here...

File: payment/src/main/java/net/optile/payment/ui/page/PaymentListPresenter.java [hunk 18]
```diff
@@ -191,15 +187,17 @@ public void onLocalizationSuccess(Localization localization) {
...
-        PaymentResult result = PaymentResult.fromThrowable(cause);
-        if (result.hasError()) {
-            handleLoadingError(result);
-        } else {
-            closeWithCanceledCode(result);
-        }
+        handleLoadingError(cause);
+    }
+
+    private PaymentResult createPaymentResult(PaymentError error) {
```
Comment (hjhill) [inline payment/src/main/java/net/optile/payment/ui/page/PaymentListPresenter.java:193]: Same code as in `PaymentListPresenter` - seems to belong on either `PaymentError` (conversion function) or `PaymentResult` (static factory).

File: payment/src/main/java/net/optile/payment/network/BaseConnection.java (hunk 16 above)
Comment (siebeprojects) [reply payment/src/main/java/net/optile/payment/network/BaseConnection.java:284]: Changing it to more specific tags again e.g. sdk_BaseConnection. In Android we are constraint to 24 chars per tag.

File: payment/src/main/java/net/optile/network/basic/BasicNetworkService.java (hunk 12 above)
Comment (siebeprojects) [reply payment/src/main/java/net/optile/network/basic/BasicNetworkService.java:151]: Changed to !=

File: payment/src/main/java/net/optile/network/basic/BasicNetworkService.java (hunk 13 above)
Comment (siebeprojects) [reply payment/src/main/java/net/optile/network/basic/BasicNetworkService.java:158]: It uses now the static method to create a PaymentResult from the PaymentError

File: payment/src/main/java/net/optile/payment/core/PaymentError.java (hunk 15 above)
Comment (siebeprojects) [reply payment/src/main/java/net/optile/payment/core/PaymentError.java:139]: Changed to isNetworkFailure in the PaymentError class

File: payment/src/main/java/net/optile/payment/ui/page/PaymentListPresenter.java (hunk 18 above)
Comment (siebeprojects) [reply payment/src/main/java/net/optile/payment/ui/page/PaymentListPresenter.java]: This has been changed to using a static method in PaymentResult to create a PaymentResult from a PaymentError

File: README.rst (hunk 2 above)
Comment (siebeprojects) [reply README.rst:221]: created a handlePaymentActivityResult method and removed the comments. 
It is better to explain which fields are provided in the README documentation than in this code sample.

File: README.rst (hunk 6 above)
Comment (siebeprojects) [reply README.rst:271]: This is fixed

File: example-demo/src/main/java/net/optile/example/demo/checkout/CheckoutPresenter.java (hunk 10 above)
Comment (siebeprojects) [reply example-demo/src/main/java/net/optile/example/demo/checkout/CheckoutPresenter.java]: Removed this comment from the example app

File: example-demo/src/main/java/net/optile/example/demo/checkout/CheckoutPresenter.java (hunk 9 above)
Comment (siebeprojects) [reply example-demo/src/main/java/net/optile/example/demo/checkout/CheckoutPresenter.java]: This is changed to hasNetworkFailure. The PaymentError has a method isNetworkFailure.

File: example-demo/src/main/java/net/optile/example/demo/summary/SummaryPresenter.java (hunk 11 above)
Comment (siebeprojects) [reply example-demo/src/main/java/net/optile/example/demo/summary/SummaryPresenter.java]: Comments have been removed

File: README.rst (hunk 1 above)
Comment (siebeprojects) [reply README.rst]: Rewrote the Payment Result chapter all together.

File: README.rst (hunk 4 above)
Comment (siebeprojects) [reply README.rst]: Rewrote the Payment Result chapter

File: README.rst (hunk 3 above)
Comment (siebeprojects) [reply README.rst]: Rewrote the PaymentResult chapter

File: README.rst (hunk 8 above)
Comment (siebeprojects) [reply README.rst]: Removed these texts all together

File: README.rst (hunk 5 above)
Comment (siebeprojects) [reply README.rst]: Rewrote the payment result chapter

File: README.rst (hunk 3 above)
Comment (siebeprojects) [reply README.rst]: Rewrote the payment result chapter

File: payment/src/main/java/net/optile/network/basic/BasicNetworkService.java [hunk 19]
```diff
@@ -90,41 +102,69 @@ public void onRedirectCanceled() {
...
+        ErrorInfo errorInfo = error.getErrorInfo();
+        PaymentResult result;
+
+        if (errorInfo != null) {
+            result = new PaymentResult(errorInfo.getInteraction(), error);
+        } else {
+            result = PaymentResult.fromPaymentError(error);
+            result.getInteraction().setCode(getErrorInteractionCode(operation));
+        }
+        presenter.onProcessPaymentResult(PaymentUI.RESULT_CODE_CANCELED, result);
```
Comment (hjhill) [inline payment/src/main/java/net/optile/network/basic/BasicNetworkService.java:156]: Still not very nice, but a bit more separation of concerns:
```suggestion
    private void handleProcessPaymentError(PaymentError error) {
        presenter.onProcessPaymentResult(PaymentUI.RESULT_CODE_CANCELED, toPaymentResult(error));
    }

    private PaymentResult toPaymentResult(PaymentError error) {
        ErrorInfo errorInfo = error.getErrorInfo();
        if (errorInfo != null) {
            return new PaymentResult(errorInfo.getInteraction(), error);
        }

        PaymentResult result = PaymentResult.fromPaymentError(error);
        result.getInteraction().setCode(getErrorInteractionCode(operation));
        return result;
```
Especially the `getInteraction().setCode(...)` bothers me - can you introduce a variant of `PaymentResult.fromPaymentError` that takes `Interaction` as the second parameter?

Comment (siebeprojects) [reply payment/src/main/java/net/optile/network/basic/BasicNetworkService.java]: Updated the BasicNetworkService and removed the static method to create a PaymentResult.
